automates/model_analysis/graph_manipulation.py

```diff
@@ -0,0 +1,538 @@
+        if prob.domain > 0:
+            if prob.dom == 1:
+                p = "".join([p, target_sym, var_string])
+            else:
+                if single_source:
+                    p = "".join([p, "(", var_string])
+                else:
+                    p = "".join([p, "^{(", str(prob.domain - 1), ")}(", var_string])
+        else:
+            p = "".join([p, "(", var_string])
+        if len(prob.cond) > 0:
+            cond_string = ",".join(prob.cond)  # prob.cond must have elements that are strings
+            cond_string = "".join(["\u007C", cond_string, ")"])
+        else:
+            cond_string = ")"
+        p = "".join([p, cond_string])
+    if s_print and start_sum:
+        p = ",".join([p, "\\right)"])
+    return p
+
+
+def c_components(g, topo):
+    """
+    Finds c-components in graph g
+    :param g: graph
+    :param topo: topological ordering
+    :return: list of c-components (each c-component is a list of nodes)
+    """
+    a = g.get_adjacency()
+    n = a.shape[0]
+    v = g.vs["name"]
+    bidirected = []
+    for i in range(0, n):
+        for j in range(0, n):
+            if a[i][j] >= 1 and a[j][i] >= 1:
+                bidirected.append(i)
+                bidirected.append(j)
+    bidirected_edges = g.es.select(_within=bidirected)
+    g_bidirected = g.subgraph_edges(bidirected_edges, delete_vertices=False)
+    subgraphs = g_bidirected.decompose()
+    cc = []
+    cc_rank = []
+    for subgraph in subgraphs:
+        nodes = ts(subgraph.vs["name"], topo)
+        cc.append(nodes)
+        rank = 0
+        for node in nodes:
+            rank = rank + topo.index(node)
+        cc_rank.append(rank)
+    (cc_sorted, rank_sorted) = list(map(list, zip(*sorted(zip(cc, cc_rank), key=lambda ab: ab[1]))))
+    cc_sorted.reverse()
+    return cc_sorted
+
+
+def parse_joint(p, v, cond, var, topo):
+    p_new = Probability()
+    p_num = copy.deepcopy(p)
+    p_num.sumset = ts(set(p.sumset) | (set(var) - set(v) - set(cond)), topo)
+    if len(cond) > 0:
+        p_den = copy.deepcopy(p)
+        p_den.sumset = ts(set(p.sumset) | (set(var) - set(cond)), topo)
+        p_new.fraction = True
+        p_new.num = copy.deepcopy(p_num)
+        p_new.den = copy.deepcopy(p_den)
+    else:
+        p_new = copy.deepcopy(p_num)
+    return p_new
+
+
+def to_names(indices, g):
+    """
+    converts vertex indices indices to vertex names
+    :param indices: list of indices
+    :param g: graph (with named nodes)
+    :return: list of vertex names
+    """
+    name_list = g.vs["name"]
+    name_sorted = [name_list[i] for i in indices]
+    return name_sorted
+
+
+def wrap_d_sep(g, x, y, z):
+    """
+    Does some quick checks before testing d-separation
+    :param g: Graph
+    :param x: nodes
+    :param y: nodes
+    :param z: nodes
+    :return: T/F if x is separated from y given z in g
+    """
+    if x == y:
+        return False
+    if len(x) == 0 or len(y) == 0:
+        return True
+    else:
```

Comment:
@aswinchester the `if` statements above are acting as _guards_. That makes the `else` here feel a little weird. If I were you I would simply put the return call after the two _if-guards_ without adding an `else`.

Reply:
This has been completed